### ISSUE TITLE
Serious bug in Tree::TreeNode#node_depth

### ISSUE THREAD

**Issue by aidansteele** [evolve75/RubyTree#42]
..and it's my fault. Refer to b3b88634eac59c8c9a7e62571768d7539ff8ae54

```ruby
require 'rubytree'
n1 = Tree::TreeNode.new 'n1'
n2 = Tree::TreeNode.new 'n2'
n3 = Tree::TreeNode.new 'n3'
n4 = Tree::TreeNode.new 'n4'
n1 << n2
n2 << n3
n3.node_depth # => 2
n4 << n1
n3.node_depth # => 2
```

The second time #node_depth is called on `n3`, it should return 3, not 2. The caching is problematic because it is only cleared when the direct parent node is changed, but is not cleared when nodes further up the tree are changed.

I suppose the best course of action for now is:

1. Revert that commit
2. Add a unit test so this regression doesn't happen again
3. (maybe) try to think of a better way to retain this optimisation

My sincere apologies for introducing this bug into your library!

**Comment by evolve75** [comment]
@aidansteele,

Thanks for catching this. Would it be possible for you to create a patch to reverse the change? 

The actual commit might be a little difficult to reverse now, as I added a few more changes, and we will have to cherry-pick the reversal if we do it via git. 

Don't worry about the regression. I did not do a public release yet.

**Comment by aidansteele** [comment]
Hi @evolve75,

I will submit a pull request fixing this and adding a unit test to catch it if it happens again. :)

**Comment by evolve75** [comment]
Thanks Aidan. Will merge and let you know!

**Comment by aidansteele** [comment]
Resolved with PR #41
